### ISSUE TITLE
Incorrect parsing of C++11 enums with underlying types.

### ISSUE THREAD

**Issue by vahancho** [universal-ctags/ctags#1178]
Hi guys,

First of all - universal-ctags is a great tool! Thanks for it.

Now about the issue.
Command line: `ctags -R -x --_xformat=[%{kind}][%{scope}][%{scopeKind}][%{access}][%{typeref}][%{signature}][%{compact}] --sort=no --extra=*-q --c++-kinds=*-zlm --fields=*-lne --language-force=C++`

#### Example 1 (good):

```
namespace Foo
{
    enum class Enum : unsigned char
    {
        A,
        B,
        C
    };
}
```
Result 1:
```
[namespace][][][-][-][-][namespace Foo]
[enum][Foo][namespace][-][-][-][enum class Enum : unsigned char]
[enumerator][Foo::Enum][enum][public][-][-][A,]
[enumerator][Foo::Enum][enum][public][-][-][B,]
[enumerator][Foo::Enum][enum][public][-][-][C ]
```

So far so good.
#### Example 2 (bad):
```
namespace Foo
{
    enum class Enum : uint8_t
    {
        A,
        B,
        C
    };
}
```
Result 2:
```
[namespace][][][-][-][-][namespace Foo]
[enum][Foo][namespace][-][-][-][enum class Enum : uint8_t]
[enumerator][Foo::uint8_t][enum][public][-][-][A,]
[enumerator][Foo::uint8_t][enum][public][-][-][B,]
[enumerator][Foo::uint8_t][enum][public][-][-][C ]
```

Note the scope field of enumerators. It's `Foo::uint8_t` instead of `Foo::Enum` as in the first example. It looks like only some basic builtin types parsed correctly.

**Comment by masatake** [comment]
Simplified test cases:

```
[jet@localhost ctags]$ cat u.cpp 
enum class Enum : unsigned char
{
A,
};
[jet@localhost ctags]$ ./ctags -o - u.cpp 
A	u.cpp	/^A,$/;"	e	enum:Enum	file:
Enum	u.cpp	/^enum class Enum : unsigned char$/;"	g	file:
[jet@localhost ctags]$ cat 8.cpp 
enum class Enum : uint8_t
{
        A,
};
[jet@localhost ctags]$ vi 8.cpp 
[jet@localhost ctags]$ ./ctags -o - 8.cpp 
A	8.cpp	/^A,$/;"	e	enum:Enum	file:
Enum	8.cpp	/^enum class Enum : uint8_t$/;"	g	file:
```

**Comment by masatake** [comment]
It seems that `:` is never considred when parsing enum block.

Comments taken from source code:
```
	/*
		Spec is:
			enum-key attr(optional) identifier(optional) enum-base(optional)
				{ enumerator-list(optional) }	(1)
			enum-key attr(optional) identifier enum-base(optional) ;
				(2)	(since C++11)
	*/
```

**Comment by masatake** [comment]
@vahancho, thank you for reporting. Maybe @pragmaware, the cxx parser maintainer may be in busy.


Here is a temporary solution:

```
diff --git a/parsers/cxx/cxx_parser.c b/parsers/cxx/cxx_parser.c
index fa3aacc..a80e4bd 100644
--- a/parsers/cxx/cxx_parser.c
+++ b/parsers/cxx/cxx_parser.c
@@ -461,6 +461,7 @@ bool cxxParserParseEnum(void)
 {
        CXX_DEBUG_ENTER();
 
+       CXXToken * pEnumName = NULL;
        unsigned int uInitialKeywordState = g_cxx.uKeywordState;
 
        if(g_cxx.pTokenChain->iCount > 1)
@@ -475,8 +476,10 @@ bool cxxParserParseEnum(void)
        */
 
        // Skip attr and class-head-name
+ retry:
        if(!cxxParserParseUpToOneOf(
                        CXXTokenTypeEOF | CXXTokenTypeSemicolon |
+                               CXXTokenTypeSingleColon |
                                CXXTokenTypeParenthesisChain | CXXTokenTypeOpeningBracket
                ))
        {
@@ -484,6 +487,16 @@ bool cxxParserParseEnum(void)
                return false;
        }
 
+       if(cxxTokenTypeIs(g_cxx.pToken,CXXTokenTypeSingleColon))
+       {
+               // check if we can extract a class name identifier
+               pEnumName = cxxTokenChainLastTokenOfType(
+                       g_cxx.pTokenChain,
+                       CXXTokenTypeIdentifier
+               );
+               goto retry;
+       }
+
        if(cxxTokenTypeIs(g_cxx.pToken,CXXTokenTypeParenthesisChain))
        {
                // probably a function declaration/prototype
@@ -521,7 +534,8 @@ bool cxxParserParseEnum(void)
        // semicolon or opening bracket
 
        // check if we can extract a class name identifier
-       CXXToken * pEnumName = cxxTokenChainLastTokenOfType(
+       if (!pEnumName)
+               pEnumName = cxxTokenChainLastTokenOfType(
                        g_cxx.pTokenChain,
                        CXXTokenTypeIdentifier
                );
```

Perfect solution may be prepared by the maintainer when he gets time.
Please, wait for awhile.

**Comment by vahancho** [comment]
@masatake, thanks for looking into this. IMO the underlying type is somehow handled by ctags, however the handling depends on which type is used. For example:

```
enum class Foo : int  // OK
enum class Foo : uint32_t // Fail. Incorrect scope of enum Foo resolved.

using MyChar = unsigned char;
enum class Foo : unsigned char // OK
enum class Foo : MyChar // Fail. Incorrect scope of enum Foo resolved.
```

**Comment by masatake** [comment]
@vahancho, yes. It depends on whether  token(s) left  of `{` is keyword or not.
int is keyword. unsigned char is keyword.

How about my patch? 
My patch tracks `:`; it picks up a token left of `:` as the name of upper scope. (if I implemented correctly.)

**Comment by vahancho** [comment]
@masatake, unfortunately I don't build ctags myself, so that I didn't try your patch yet.
I would extend the list of keywords with new type names. For example, [here](http://en.cppreference.com/w/cpp/types/integer) is the list of integer types introduced by C++11.
From the other hand that's not enough, as a type used can be an alias of another type, as shown in my last example. Probably your patch will solve that case too.

**Comment by masatake** [comment]
> @masatake, unfortunately I don't build ctags myself, so that I didn't try your patch yet.
I see.

**Comment by pragmaware** [comment]
Fixed via #1184.

**Comment by vahancho** [comment]
Thanks for the quick fix. Great job!
